dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -882,6 +902,347 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+                        + " Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+            }
+        }
+    }
+
+
+    /**
+     * Check if a Zenodo DOI already exists for the workflow version
+     * @param workflowVersion workflow version
+     */
+    private void checkForExistingDOIForWorkflowVersion(WorkflowVersion workflowVersion) {
+        String workflowVersionDoiURL = workflowVersion.getDoiURL();
+        if (workflowVersionDoiURL != null && !workflowVersionDoiURL.isEmpty()) {
+            LOG.error("Workflow version " + workflowVersion.getName() + " already has DOI " + workflowVersionDoiURL
+                    + ". Dockstore will only create one DOI per version.");
+            throw new CustomWebApplicationException("Workflow version " + workflowVersion.getName() + " already has DOI "
+                    + workflowVersionDoiURL + ". Dockstore will only create one DOI per version.", HttpStatus.SC_METHOD_NOT_ALLOWED);
+        }
+    }
+
+    /**
+     * Get an existing Zenodo DOI for the workflow if one exists
+     * otherwise return null
+     * @param workflow workflow
+     */
+    private String getAnExistingDOIForWorkflow(Workflow workflow) {
+        // Find out if this workflow already has at least one
+        // version that has been assigned a DOI
+        // If a version DOI exists, we will create another version DOI
+        // instead of creating a new workflow concept DOI and version DOI
+        // Get the ID of one of the workflow version DOIs
+        // because Zenodo requires that we use it to create the next
+        // workflow version DOI
+        String latestWorkflowVersionDOIURL = null;
+        Set<WorkflowVersion> setOfWorkflowVersions = workflow.getWorkflowVersions();
+        Iterator iter = setOfWorkflowVersions.iterator();
+        while (iter.hasNext()) {
+            WorkflowVersion myWorkflowVersion = ((WorkflowVersion)iter.next());
+            latestWorkflowVersionDOIURL = myWorkflowVersion.getDoiURL();
+            if (latestWorkflowVersionDOIURL != null && !latestWorkflowVersionDOIURL.isEmpty()) {
+                break;
+            }
+        }
+        return latestWorkflowVersionDOIURL;
+    }
+
+    /**
+     * Check if a Zenodo DOI already exists for the workflow version
+     * @param depositApi Zenodo API for working with depositions
+     * @param depositMetadata Metadata for the workflow version
+     * @param depositionID Zenodo's ID for the deposition
+     */
+    private void putDepositionOnZenodo(DepositsApi depositApi, DepositMetadata depositMetadata, int depositionID) {
+        NestedDepositMetadata nestedDepositMetadata = new NestedDepositMetadata();
+        nestedDepositMetadata.setMetadata(depositMetadata);
+        try {
+            depositApi.putDeposit(depositionID, nestedDepositMetadata);
+        } catch (ApiException e) {
+            LOG.error("Could not put deposition metadata on Zenodo. Error is " + e.getMessage());
```

Comment:
Please log the exception as a second parameter.

Reply:
done